applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.cpp

```diff
@@ -288,220 +261,53 @@ namespace Kratos
+		mpCoordinateTransformation->FinalizeNonLinearIteration(rCurrentProcessInfo);
 
-		const GeometryType & geom = this->GetGeometry();
-		const Matrix & shapeFunctionsValues = geom.ShapeFunctionsValues(GetIntegrationMethod());
-		for (SizeType i = 0; i < mSections.size(); i++)
-			mSections[i]->FinalizeNonLinearIteration(GetProperties(), geom, row(shapeFunctionsValues, i), CurrentProcessInfo);
+		BaseFinalizeNonLinearIteration(rCurrentProcessInfo);
 	}
 
-	void ShellThickElement3D3N::InitializeSolutionStep(ProcessInfo& CurrentProcessInfo)
+	void ShellThickElement3D3N::InitializeSolutionStep(ProcessInfo& rCurrentProcessInfo)
 	{
-		const PropertiesType& props = GetProperties();
-		const GeometryType & geom = GetGeometry();
-		const Matrix & shapeFunctionsValues = geom.ShapeFunctionsValues(GetIntegrationMethod());
+		BaseInitializeSolutionStep(rCurrentProcessInfo);
 
-		for (SizeType i = 0; i < mSections.size(); i++)
-			mSections[i]->InitializeSolutionStep(props, geom, row(shapeFunctionsValues, i), CurrentProcessInfo);
-
-		mpCoordinateTransformation->InitializeSolutionStep(CurrentProcessInfo);
+		mpCoordinateTransformation->InitializeSolutionStep(rCurrentProcessInfo);
 	}
 
-	void ShellThickElement3D3N::FinalizeSolutionStep(ProcessInfo& CurrentProcessInfo)
+	void ShellThickElement3D3N::FinalizeSolutionStep(ProcessInfo& rCurrentProcessInfo)
 	{
-		const PropertiesType& props = GetProperties();
-		const GeometryType& geom = GetGeometry();
-		const Matrix & shapeFunctionsValues = geom.ShapeFunctionsValues(GetIntegrationMethod());
+		BaseFinalizeSolutionStep(rCurrentProcessInfo);
 
-		for (SizeType i = 0; i < mSections.size(); i++)
-			mSections[i]->FinalizeSolutionStep(props, geom, row(shapeFunctionsValues, i), CurrentProcessInfo);
-
-		mpCoordinateTransformation->FinalizeSolutionStep(CurrentProcessInfo);
+		mpCoordinateTransformation->FinalizeSolutionStep(rCurrentProcessInfo);
 	}
 
 	void ShellThickElement3D3N::CalculateMassMatrix(MatrixType& rMassMatrix, ProcessInfo& rCurrentProcessInfo)
 	{
-		if ((rMassMatrix.size1() != OPT_NUM_DOFS) || (rMassMatrix.size2() != OPT_NUM_DOFS))
-			rMassMatrix.resize(OPT_NUM_DOFS, OPT_NUM_DOFS, false);
-		noalias(rMassMatrix) = ZeroMatrix(OPT_NUM_DOFS, OPT_NUM_DOFS);
+		if ((rMassMatrix.size1() != 18) || (rMassMatrix.size2() != 18))
+			rMassMatrix.resize(18, 18, false);
+		noalias(rMassMatrix) = ZeroMatrix(18, 18);
 
 		// Compute the local coordinate system.
 		ShellT3_LocalCoordinateSystem referenceCoordinateSystem(
 			mpCoordinateTransformation->CreateReferenceCoordinateSystem());
 
 		// Average mass per unit area over the whole element
 		double av_mass_per_unit_area = 0.0;
-		for (std::size_t i = 0; i < OPT_NUM_GP; i++)
+
+        const SizeType num_gps = GetNumberOfGPs();
+
+		for (SizeType i = 0; i < num_gps; i++)
 			av_mass_per_unit_area += mSections[i]->CalculateMassPerUnitArea();
-		av_mass_per_unit_area /= double(OPT_NUM_GP);
+		av_mass_per_unit_area /= double(num_gps);
```

Comment:
minor (and don't need to change) however if you cast, do it by 

        static_cast<double>(num_gps);

if you take a look in stack overflow there are long discussions about the why. In my opinion a convincing reason is that it makes evident where the casting is done by a simple search in the archive. As i am telling it is of minor importance though